syntax = "proto3";
package chat;

enum StatusCode {
    STATUS_UNKNOWN = 0;
    STATUS_SUCCESS = 1;
    STATUS_FAILURE = 2;
    STATUS_UNAUTHORIZED = 3;
    STATUS_NOT_FOUND = 4;
}

message Status {
    StatusCode code = 1;
    optional string message = 2;
}

enum UserRights {
    ADMIN = 0;
    OWNER = 1;
    MODERATOR = 2;
    REGULAR = 3;
}

message UserInfo {
    int32 user_id = 1;
    string user_name = 2;
    optional UserRights user_room_rights = 3;
}

message MessageInfo {
    string from = 1;
    string message = 2;
    uint64 timestamp = 3;
}

message RoomInfo {
    int32 room_id = 1;
    string room_name = 2;
    optional UserInfo owner = 3;
}

message GenericError {
    Status status = 1;
}

message ServerHello {
    string message = 1;
}

message AuthRequest {
    string username = 1;
    string password = 2;
}
message AuthResponse {
    Status status = 1;
}

message RegisterRequest {
    string username = 1;
    string password = 2;
}
message RegisterResponse {
    Status status = 1;
}

message SendMessageRequest {
    string message = 1;
}
message SendMessageResponse {
    Status status = 1;
}

message JoinRoomRequest {
    int32 room_id = 1;
}
message JoinRoomResponse {
    Status status = 1;
    repeated UserInfo users = 2;
}

message LeaveRoomRequest {
}
message LeaveRoomResponse {
    Status status = 1;
}

message GetRoomsRequest {
}
message GetRoomsResponse {
    Status status = 1;
    repeated RoomInfo rooms = 2;
}

message CreateRoomRequest {
    string room_name = 1;
}
message CreateRoomResponse {
    Status status = 1;
    RoomInfo info = 2;
}

message RoomMessage {
    MessageInfo message = 1;
}

message GetMessagesRequest{
    uint32 limit = 1;
    uint32 offset = 2;
}
message GetMessagesResponse{
    Status status = 1;
    repeated MessageInfo message = 2;
}

message Envelope {
    oneof payload {
        ServerHello server_hello = 1;
        AuthRequest auth_request = 2;
        AuthResponse auth_response = 3;
        RegisterRequest register_request = 4;
        RegisterResponse register_response = 5;
        SendMessageRequest send_message_request = 6;
        SendMessageResponse send_message_response = 7;
        JoinRoomRequest join_room_request = 8;
        JoinRoomResponse join_room_response = 9;
        LeaveRoomRequest leave_room_request = 10;
        LeaveRoomResponse leave_room_response = 11;
        GetRoomsRequest get_rooms_request = 12;
        GetRoomsResponse get_rooms_response = 13;
        CreateRoomRequest create_room_request = 14;
        CreateRoomResponse create_room_response = 15;
        RoomMessage room_message = 16;
        GetMessagesRequest get_messages_request = 17;
        GetMessagesResponse get_messages_response = 18;
        GenericError generic_error = 19;
    }
}
